suite/tests/client-interface/drx-scattergather-instrumentation-fault.dll.c

```diff
@@ -0,0 +1,279 @@
+
+    if (drreg_unreserve_register(drcontext, ilist, where, reg_ptr) != DRREG_SUCCESS)
+        DR_ASSERT(false);
+}
+
+static void
+handle_post_instruction(void *drcontext, instrlist_t *ilist, instr_t *where,
+                        app_pc app_addr)
+{
+    int i;
+    instr_t *prev_instr = instr_get_prev_app(where);
+    bool seen_memref = false;
+
+    for (i = 0; i < instr_num_dsts(prev_instr); ++i) {
+        const opnd_t dst = instr_get_dst(prev_instr, i);
+        if (opnd_is_memory_reference(dst)) {
+            if (seen_memref) {
+                DR_ASSERT_MSG(false, "Found inst with multiple memory destinations");
+                break;
+            }
+
+            seen_memref = true;
+            insert_faulting_store(drcontext, ilist, where, app_addr);
+        }
+    }
+    for (i = 0; i < instr_num_srcs(prev_instr); ++i) {
+        const opnd_t src = instr_get_src(prev_instr, i);
+        if (opnd_is_memory_reference(src)) {
+            if (seen_memref) {
+                DR_ASSERT_MSG(false, "Found inst with multiple memory destinations");
+                break;
+            }
+
+            seen_memref = true;
+            insert_faulting_store(drcontext, ilist, where, app_addr);
+        }
+    }
+}
+
+static dr_emit_flags_t
+event_app_analysis(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
+                   bool translating, void **user_data)
+{
+    instru_data_t *data = (instru_data_t *)dr_thread_alloc(drcontext, sizeof(*data));
+    data->app_addr = 0;
+    data->is_scatter_gather = false;
+    data->do_post_instruction_instrumentation = false;
+
+    /* Detect whether this bb is a scatter/gather expansion. */
+    for (instr_t *instr = instrlist_first(bb); instr != NULL;
+         instr = instr_get_next(instr)) {
+        if (drmgr_is_emulation_start(instr)) {
+            emulated_instr_t emulated_instr;
+            emulated_instr.size = sizeof(emulated_instr);
+            if (drmgr_get_emulated_instr_data(instr, &emulated_instr)) {
+                data->is_scatter_gather = instr_is_scatter(emulated_instr.instr) ||
+                    instr_is_gather(emulated_instr.instr);
+                if (data->is_scatter_gather) {
+                    data->app_addr = instr_get_app_pc(emulated_instr.instr);
+                    break;
+                }
+            }
+        }
+    }
+
+    *user_data = (void *)data;
+    return DR_EMIT_DEFAULT;
+}
+
+static dr_emit_flags_t
+event_app_instruction(void *drcontext, void *tag, instrlist_t *bb, instr_t *where,
+                      bool for_trace, bool translating, void *user_data)
+{
+    int i;
+    bool seen_memref = false;
+    instru_data_t *data = (instru_data_t *)user_data;
+
+    /* If the previous instruction was a scatter/gather expansion load/store,
+     * we insert another faulting store after it.
+     */
+    if (data->do_post_instruction_instrumentation) {
+        handle_post_instruction(drcontext, bb, where, data->app_addr);
```

Comment:
Why is this post instr instrumentation needed? Does it test some scenario that the other part of the test doesn't? If not we can simplify this client by removing the post instr instrumentation or else maybe add a comment saying why it is needed.